### PR TITLE
Decoder key export fixes

### LINKED ISSUE #21493: OpenSSL 3: x25519 a decode from and then encode to a pem file corrupts the key if fips+base provider is used
This issue ticket is for the x25519 crypto version of another issue with ed25519 crypto: <https://github.com/openssl/openssl/issues/20758#issuecomment-1639668646>. As the issue with the ed25519 has a difficulty to reproduce with OpenSSL master branch, and includes another issue. So, I created this issue ticket not to mix the 2 issues. And as we can reproduce this issue with the x25519 crypto file on the master branch (OpenSSL 3.2.0.dev), it's more convenient to debug and fix it.

I checked this issue on the current latest master branch `8c34367e434c6b9555f21cc4fc77a18d6ef84a85`. The x25519 crypto is approved in the FIPS case in the master branch as a reference.

```
$ grep PROV_NAMES_X25519 providers/fips/fipsprov.c
    { PROV_NAMES_X25519, FIPS_DEFAULT_PROPERTIES, ossl_x25519_keyexch_functions },
    { PROV_NAMES_X25519, FIPS_DEFAULT_PROPERTIES, ossl_x25519_keymgmt_functions,
```

## Summary

When reading a x25519 crypto public key pem file, then encoding and decoding it in FIPS case with "base" and "fips" providers, the decoded text is different from the original one in only FIPS case.

## Reproducing steps

I prepared the [reproducing program](https://github.com/junaruga/report-openssl-fips-base-decoding-corruption) whose repository name was previously "report-openssl-fips-ed25519" and used in the #20758.

Prepare the x25519 public key pem file.

```
$ cat x25519_pub.pem
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----
```

Compile the reproducing program with the targeting OpenSSL.

```
$ gcc \
  -I /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/include \
  -L /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/lib \
  -o reproducer reproducer.c -lcrypto
```

### Non-FIPS case

The decoded text is the same with the original PEM file. This is ok.

```
$ OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/lib \
  ./reproducer x25519_pub.pem
[DEBUG] Loaded providers:
  default
[DEBUG] FIPS mode enabled: 0
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0x5707a0
[DEBUG] It's held by the provider default
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----
```

### FIPS case

The decode text is different from the original PEM file. This is not ok.

```
$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-8c34367e43/lib \
  ./reproducer x25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0xd7fde0
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
-----END PUBLIC KEY----
```

### PR DESCRIPTION
**Pull request by t8m**
This corrects two serious problems - one in decoders where we try to export/import things with 0 selection which is not meaningful and the second in the actual keymgmt export functions which instead of failing on such export call, they produce bogus (empty) keys.

Fixes #21493

### CODE REVIEW

**Comment by t8m** [comment]
I'll add testcases next.

**Review by paulidale — APPROVED** [review]

**Comment by junaruga** [comment]
## master branch

I tested this PR with [my reproducer](https://github.com/junaruga/report-openssl-fips-base-decoding-corruption), the x25519 public key PEM file, and the OpenSSL on the current latest master branch 06a0d40322e96dbba816b35f82226871f635ec5a plus the commits in this PR. And I can see the encoded and decoded text is same with the original one!

### master branch plus this PR's commits

The encoded and decode text is same with the original one. It's okay.

```
$ cat x25519_pub.pem
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----

$ gcc \
  -I /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322-plus-pr-21519/include \
  -L /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322-plus-pr-21519/lib \
  -o reproducer_pr_21519 reproducer.c -lcrypto

$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322-plus-pr-21519/ssl \ 
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322-plus-pr-21519/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322-plus-pr-21519/lib \
  ./reproducer_pr_21519 x25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0xbe9e40
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----
```

### master branch without this PR's commits

And the encoded and decoded text is different from the original one on the current master branch 06a0d40322e96dbba816b35f82226871f635ec5a without this PR's commits. This is the expected behavior. I checked it just in case.

```
$ gcc \
  -I /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322/include \
  -L /home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322/lib \
  -o reproducer reproducer.c -lcrypto

$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.2.0.dev-fips-debug-06a0d40322/lib \
  ./reproducer x25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0x83adf0
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
-----END PUBLIC KEY-----
```

**Comment by junaruga** [comment]
I tested this PR on the lastet openssl-3.1 and openssl-3.1 branches too. It looks good to me!


## openssl-3.1 branch plus this PR's commits

```
$ gcc \
  -I /home/jaruga/.local/openssl-3.1.2.dev-fips-debug-cead7e5adf-plus-pr-21519/include \
  -L /home/jaruga/.local/openssl-3.1.2.dev-fips-debug-cead7e5adf-plus-pr-21519/lib \
  -o reproducer_pr_21519_312 reproducer.c -lcrypto
```

```
$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.1.2.dev-fips-debug-cead7e5adf-plus-pr-21519/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.1.2.dev-fips-debug-cead7e5adf-plus-pr-21519/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.1.2.dev-fips-debug-cead7e5adf-plus-pr-21519/lib \
  ./reproducer x25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0x1c6fe40
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----
```

## openssl-3.0 branch plus this PR's commits

I tested both x25519 and ed25519 cases.

```
$ gcc \
  -I /home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/include \
  -L /home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/lib \
  -o reproducer_pr_21519_3010 reproducer.c -lcrypto
```

### crypto: x25519

```
$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/lib \
  ./reproducer_pr_21519_3010 x25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0x218c550
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VuAyEA3p7bfXt9wbTTW2HC7OQ1Nz+DQ8hbeGdNrfx+FG+IK08=
-----END PUBLIC KEY-----
```

### crypto: ed25519

```
$ cat ed25519_pub.pem
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VwAyEAPUAXw+hDiVqStwqnTRt+vJyYLM8uxJaMwM1V8Sr0Zgw=
-----END PUBLIC KEY-----
```

```
$ OPENSSL_CONF=$(pwd)/openssl_fips.cnf \
  OPENSSL_CONF_INCLUDE=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/ssl \
  OPENSSL_MODULES=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/lib/ossl-modules \
  LD_LIBRARY_PATH=/home/jaruga/.local/openssl-3.0.10.dev-fips-debug-6bcb6d297e-plus-pr-21519/lib \
  ./reproducer_pr_21519_3010 ed25519_pub.pem
[DEBUG] Loaded providers:
  fips
  base
[DEBUG] FIPS mode enabled: 1
[DEBUG] data_size: 113
[DEBUG] OSSL_DECODER_from_bio PEM 2 failed.
[DEBUG] Got a pkey! 0xf7c550
[DEBUG] It's held by the provider fips
[DEBUG] ossl_membio2str buf->data:
-----BEGIN PUBLIC KEY-----
MCowBQYDK2VwAyEAPUAXw+hDiVqStwqnTRt+vJyYLM8uxJaMwM1V8Sr0Zgw=
-----END PUBLIC KEY-----
```

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 1]
```diff
@@ -222,11 +222,14 @@ static int dh_export(void *keydata, int selection, OSSL_CALLBACK *param_cb,
...
 
+    if ((selection & DH_POSSIBLE_SELECTIONS) == 0)
+        return 0;
+
     tmpl = OSSL_PARAM_BLD_new();
     if (tmpl == NULL)
         return 0;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_ALL_PARAMETERS) != 0)
+    if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0)
```
[the remark this answers is not in the document]
Comment (t8m) [reply providers/implementations/keymgmt/dh_kmgmt.c:232]: I'll drop this change.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale please reconfirm
@mattcaswell fixup pushed

**Review by paulidale — APPROVED** [review]

**Comment by junaruga** [comment]
Thank you!

**Comment by junaruga** [comment]
> Merged to `master`, `openssl-3.1` and `openssl-3.0`. Thank you!

I just put the note that this PR's commits were merged as 3 commits to master, openssl-3.1 and openssl-3.0 branches for someone who is interested in it.

* master branch
  * https://github.com/openssl/openssl/commit/1ae4678cebaa13604c0f31bdf2c64cd28bdaf287
  * https://github.com/openssl/openssl/commit/2acb0d363c0032b5b97c4f6596609f40bd7d842f
  * https://github.com/openssl/openssl/commit/4c50610bdadbcf7aa6bbd968df67b8874234677b
* openssl-3.1
  * https://github.com/openssl/openssl/commit/8865d7c4e8f9afac969fc927f2b24ee6d03868bd
  * https://github.com/openssl/openssl/commit/137ba0567417441cd8b3d43cf23e27d73f7a7684
  * https://github.com/openssl/openssl/commit/42f32b460646d8d4ef1e8927ec976716f02a093a
* openssl-3.0
  * https://github.com/openssl/openssl/commit/b5edc8d02ce0c11f39529b0fc95ae64c1c9f6c70
  * https://github.com/openssl/openssl/commit/52080c8bafcfd445d0d388cc93ad5d1f0d51bf39
  * https://github.com/openssl/openssl/commit/2d79662b5febe565d39861338a5d9067a2df641f

**Comment by levitte** [comment]
> ... with 0 selection which is not meaningful

Actually, from `man OSSL_DECODER_CTX_new_for_pkey`:

> The search of decoder implementations can also be limited with _keytype_ and _selection_, which specifies the expected resulting keytype and contents.  NULL and zero are valid and signify that the decoder implementations will find out the keytype and key contents on their own from the input they get.

So, the selection 0 means "get me whatever you can", and leaves it to the application to discover (through probing) what it actually got.

**Comment by t8m** [comment]
> > ... with 0 selection which is not meaningful
> 
> Actually, from `man OSSL_DECODER_CTX_new_for_pkey`:
> 
> > The search of decoder implementations can also be limited with _keytype_ and _selection_, which specifies the expected resulting keytype and contents.  NULL and zero are valid and signify that the decoder implementations will find out the keytype and key contents on their own from the input they get.
> 
> So, the selection 0 means "get me whatever you can", and leaves it to the application to discover (through probing) what it actually got.

Yes, it is meaningful for OSSL_DECODER_CTX_new_for_pkey() but not for (at least in the current implementation of providers) for keymgmt import and export functions.

**Pull request closed without merging**